src/iam/iam-base.ts

```diff
@@ -245,38 +255,97 @@ export class IAMBase {
-      this._walletConnectProvider.wc.on("display_uri", (err, payload) => {
-        // uri is expected to be WalletConnect Standard URI https://eips.ethereum.org/EIPS/eip-1328
-        const wcUri = payload.params[0];
-
-        const encoded = encodeURIComponent(wcUri);
-        const hasQueryString = this._ewKeyManagerUrl.includes("?");
-        const url = `${this._ewKeyManagerUrl}${hasQueryString ? "&" : "?"}uri=${encoded}`;
-        window.open(url, "_blank");
+      const showQRCode = !(walletProvider === WalletProvider.EwKeyManager);
+
+      this._walletConnectProvider = new WalletConnectProvider({
+        rpc: {
+          [this._connectionOptions.chainId]: this._connectionOptions.rpcUrl
+        },
+        infuraId: this._connectionOptions.infuraId,
+        bridge: this._connectionOptions.bridgeUrl,
+        qrcode: showQRCode
       });
+
+      if (walletProvider === WalletProvider.EwKeyManager) {
+        this._walletConnectProvider.wc.on("display_uri", (err, payload) => {
+          // uri is expected to be WalletConnect Standard URI https://eips.ethereum.org/EIPS/eip-1328
+          const wcUri = payload.params[0];
+
+          const encoded = encodeURIComponent(wcUri);
+          const hasQueryString = this._ewKeyManagerUrl.includes("?");
+          const url = `${this._ewKeyManagerUrl}${hasQueryString ? "&" : "?"}uri=${encoded}`;
+          window.open(url, "_blank");
+        });
+      }
+
+      await this._walletConnectProvider.enable();
+
+      this._walletConnectProvider.wc.on("session_update", () => {
+        this.closeConnection();
+      });
+
+      this._walletConnectProvider.wc.on("disconnect", () => {
+        this.closeConnection();
+      });
+
+      this._signer = new providers.Web3Provider(this._walletConnectProvider).getSigner();
+      this._providerType = walletProvider;
+      return;
     }
+    throw new Error(ERROR_MESSAGES.WALLET_TYPE_NOT_PROVIDED);
+  }
 
-    await this._walletConnectProvider.enable();
-    this._signer = new providers.Web3Provider(this._walletConnectProvider).getSigner();
+  private storeSession() {
+    if (this._runningInBrowser && window && this._didSigner) {
+      sessionStorage.setItem(WALLET_PROVIDER, this._providerType as string);
+      sessionStorage.setItem(PUBLIC_KEY, this._didSigner.publicKey);
+    }
+  }
+
+  protected clearSession() {
+    if (this._runningInBrowser && window) {
+      sessionStorage.removeItem(WALLET_PROVIDER);
+      sessionStorage.removeItem(PUBLIC_KEY);
+    }
+  }
+
+  /**
+   * Close connection to wallet
+   * @description closes the connection between dApp and the wallet
+   *
+   */
+
+  async closeConnection() {
+    if (this._walletConnectProvider) {
+      await this._walletConnectProvider.close();
+    }
+    this.clearSession();
+    this._did = undefined;
+    this._address = undefined;
+    this._signer = undefined;
+    if (this._runningInBrowser) {
+      location.reload();
```

Comment:
I'm a little hesitant to be having the client-lib reloading apps which are depending on it. Not sure if it's in the scope of this story, but maybe it would be better to have the client-lib be an eventEmitter and emit a "closed" event and then apps can reload if they want? I think we might have discussed doing something like this at some point recently.

Reply:
Good point 👍  I'll remove this for now, and afterwards we can add this `on` method